File: src/main/java/com/leggo/cooperativa/infrastructure/repositories/mongodb/ProductMongoRepository.java
package com.leggo.cooperativa.infrastructure.repositories.mongodb;

import com.leggo.cooperativa.domain.model.product.Product;
import com.leggo.cooperativa.domain.model.product.ProductId;
import com.leggo.cooperativa.domain.repositories.ProductRepository;
import lombok.RequiredArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@RequiredArgsConstructor
public class ProductMongoRepository implements ProductRepository
{
    private final ProductMongo mongo;

    @Override
    public void addProduct(Product product)
    {
        Map<LocalDate, BigDecimal>pricesByDay = product.getMarketRates().getPricesByDay().entrySet().stream()
            .collect(Collectors.toMap(Map.Entry::getKey, it -> it.getValue().getAmount()));

        ProductEntity.MarketRateEntity marketRateEntity = ProductEntity.MarketRateEntity.builder()
            .initialPricePerKilogram(product.getMarketRates().getInitialPricePerKilogram().getAmount())
            .pricesByDay(pricesByDay).build();

        ProductEntity entity = ProductEntity.builder()
            .productId(product.getProductId().getId())
            .name(product.getName())
            .kilogramsPerHectare(product.getKilogramsPerHectare().getAmount())
            .marketRates(marketRateEntity).build();

        mongo.save(entity);
    }

    @Override
    public Optional<Product> findProductById(ProductId productId)
    {
        return Optional.empty();
    }

    @Override
    public void updateProduct(Product product)
    {

    }
}

File: src/main/java/com/leggo/cooperativa/infrastructure/repositories/mongodb/ProductMongoRepository.kt
package com.leggo.cooperativa.infrastructure.repositories.mongodb

import com.leggo.cooperativa.domain.model.common.KilogramsPerHectare
import com.leggo.cooperativa.domain.model.common.PricePerKilogram
import com.leggo.cooperativa.domain.model.product.MarketRate
import com.leggo.cooperativa.domain.model.product.Product
import com.leggo.cooperativa.domain.model.product.ProductId
import com.leggo.cooperativa.domain.model.product.ProductType
import com.leggo.cooperativa.domain.repositories.ProductRepository
import com.leggo.cooperativa.infrastructure.repositories.mongodb.entities.MarketRateEntity
import com.leggo.cooperativa.infrastructure.repositories.mongodb.entities.ProductEntity
import java.util.Optional
import java.util.TreeMap

class ProductMongoRepository
(
    private val mongo: ProductMongo
) : ProductRepository
{
    override fun addProduct(product: Product)
    {
        mongo.save(product.toEntity())
    }

    override fun findProductById(productId: ProductId): Optional<Product>
    {
        val domain = mongo.queryByProductId(productId.id)?.toDomain()
        return Optional.ofNullable(domain)
    }

    override fun updateProduct(product: Product)
    {
        mongo.save(product.toEntity())
    }

    // ADAPTERS
    //--------------------------------------------------------------------------------------------------------

    private fun Product.toEntity(): ProductEntity
    {
        val marketRatesEntity = MarketRateEntity(
            initialPricePerKilogram = this.marketRates.initialPricePerKilogram.amount,
            pricesByDay = this.marketRates.pricesByDay.mapValues { it.value.amount }.toMap() )

        return ProductEntity(
            productId = this.productId.id,
            name = this.name,
            kilogramsPerHectare = this.kilogramsPerHectare.amount,
            marketRates = marketRatesEntity,
            productType = this.productType.name)
    }

    private fun ProductEntity.toDomain(): Product
    {
        val pricesByDay = this.marketRates.pricesByDay.mapValues { PricePerKilogram.of(it.value) }.let { TreeMap(it) }
        val initialPricePerKilogram = PricePerKilogram.of(this.marketRates.initialPricePerKilogram)
        val marketRates = MarketRate(pricesByDay, initialPricePerKilogram)

        return Product(
            ProductId.of(this.productId),
            this.name,
            KilogramsPerHectare.of(this.kilogramsPerHectare),
            marketRates,
            ProductType.valueOf(this.productType) )
    }
}
